libs/librepcb/common/uuid.cpp

```diff
@@ -34,19 +34,59 @@ namespace librepcb {
  *  Static Methods
  ******************************************************************************/
 
+inline bool isLowerHex(const QChar chr) noexcept {
+  return (chr >= QChar('0') && chr <= QChar('9')) || (chr >= QChar('a') && chr <= QChar('f'));
```

Comment:
- This method should be either marked as `static`, or (even better) moved to inside the class (static too, although then `static` has a different meaning). Otherwise the method pollutes the global namespace since it is not limited to this translation unit.
- The parameter could probably be passed as a const *reference* (const *value* doesn't make that much sense).
- Maybe it would be even (slightly) more efficient when avoiding the creation of the temporary `QChar` objects? I mean, the passed `QChar` could be converted to `char` with `toLatin1()` to allow comparing them directly with e.g. `'a'`.
- I suspect this line is not formatted with clang-format :wink:

Reply:
> This method should be either marked as static, or (even better) moved to inside the class

I replaced it with a lambda function instead, then it's even confined to the method. According to callgrind (if I used and interpreted it correctly) the instruction count for a release build did not change, so both should be optimized the same way.

> The parameter could probably be passed as a const reference (const value doesn't make that much sense).

Done as well, doesn't change the generated code. (I'm not sure how this is implemented in C++, but in Rust, passing a reference would be 64 bits (pointer size) while passing a char would be 8 bits, so for primitive types pass-by-value is preferred. In the worst case it's the same performance as a pointer, if padding to pointer size is used.)

> the passed QChar could be converted to char with toLatin1() to allow comparing them directly with e.g. 'a'.

I tested it:

```diff
   // Helper function
   auto isLowerHex = [](const QChar& chr) {
-    return (chr >= QChar('0') && chr <= QChar('9'))
-        || (chr >= QChar('a') && chr <= QChar('f'));
+    char c = chr.toLatin1();
+    return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f');
   };
```

However, it seems that the instruction count per call went up from 768 to 803. Probably because of the overhead of the `toLatin1` function call and the conditional inside. I'd guess that the compiler can optimize `QChar('a')` quite well.

> I suspect this line is not formatted with clang-format wink

Should be formatted now.

Reply:
> I replaced it with a lambda function instead, then it's even confined to the method.

Thanks, looks good! :+1:

> I'm not sure how this is implemented in C++, but in Rust, passing a reference would be 64 bits (pointer size) while passing a char would be 8 bits, so for primitive types pass-by-value is preferred. In the worst case it's the same performance as a pointer, if padding to pointer size is used.

For POD types I agree. But `QChar` is not POD so we would have to look at its implementation to know its size (if it's not documented by Qt). In addition, passing an object by value will call its copy constructor which could be much more expensive than just passing a reference. Of course, this overhead might be optimized away if the copy constructor is trivial...

Reply:
> In addition, passing an object by value will call its copy constructor which could be much more expensive than just passing a reference.

Ah, thanks, good to know!